src/structures/ThreadChannel.js

```diff
@@ -0,0 +1,394 @@
+
+    /**
+     * The timestamp the thread was last archived or unarchived at
+     * @type {?number}
+     */
+    this.archiveTimestamp = data.thread_metadata.archive_timestamp
+      ? new Date(data.thread_metadata.archive_timestamp).getTime()
+      : null;
+
+    /**
+     * The approximate count of messages in this thread
+     * <info>This value will not count above 50 even when there are more than 50 messages
+     * If you need an approximate value higher than this, use ThreadChannel#messages.cache.size</info>
+     * @type {number}
+     */
+    this.messageCount = data.message_count;
+
+    /**
+     * The approximate count of users in this thread
+     * <info>This value will not count above 50 even when there are more than 50 members</info>
+     * @type {number}
+     */
+    this.memberCount = data.member_count;
+
+    if (data.member && this.client.user?.id) this.members.add({ user_id: this.client.user.id, ...data.member });
+    if (data.messages) for (const message of data.messages) this.messages.add(message);
+  }
+
+  /**
+   * A collection of the guild member objects for each of this thread's members
+   * @type {Collection<Snowflake, GuildMember>}
+   * @readonly
+   */
+  get guildMembers() {
+    return this.members.cache.mapValues(member => member.guildMember);
+  }
+
+  /**
+   * The time the thread was last archived or unarchived at
+   * @type {?Date}
+   * @readonly
+   */
+  get archivedAt() {
+    return this.archiveTimestamp ? new Date(this.archiveTimestamp) : null;
+  }
+
+  /**
+   * The parent channel of this thread
+   * @type {?(NewsChannel|TextChannel)}
+   * @readonly
+   */
+  get parent() {
+    return this.guild.channels.resolve(this.parentID);
+  }
+
+  /**
+   * Adds a member to this thread.
+   * @param {UserResolvable} member The member to add
+   * @param {string} [reason] The reason for adding this member
+   * @returns {Promise<Snowflake>}
+   */
+  addMember(member, reason) {
+    const id = member === '@me' ? member : this.client.users.resolveID(member);
+    if (!id) return Promise.reject(new TypeError('INVALID_TYPE', 'member', 'UserResolvable'));
+    return this.client.api
+      .channels(this.id, 'thread-members', id)
+      .put({ reason })
+      .then(() => id);
+  }
+
+  /**
+   * Makes the client user join the thread.
+   * @returns {Promise<ThreadChannel>}
+   */
+  join() {
+    return this.addMember('@me').then(() => this);
+  }
+
+  /**
+   * Makes the client user leave the thread.
+   * @returns {Promise<ThreadChannel>}
+   */
+  leave() {
+    return this.members.remove('@me').then(() => this);
+  }
+
+  /**
+   * Gets the overall set of permissions for a member or role in this threads' parent, taking into account overwrites.
+   * @param {GuildMemberResolvable|RoleResolvable} memberOrRole The member or role to obtain the overall permissions for
+   * @returns {?Readonly<Permissions>}
+   */
+  permissionsFor(memberOrRole) {
+    return this.parent?.permissionsFor(memberOrRole);
```

Comment:
```suggestion
    return this.parent?.permissionsFor(memberOrRole) ?? null;
```

`x?.y();` could return undefined